cmd/kubeadm/app/phases/controlplane/manifests.go

```diff
@@ -48,9 +48,12 @@ func CreateInitStaticPodManifestFiles(manifestDir, patchesDir string, cfg *kubea
 
 // GetStaticPodSpecs returns all staticPodSpecs actualized to the context of the current configuration
 // NB. this method holds the information about how kubeadm creates static pod manifests.
-func GetStaticPodSpecs(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmapi.APIEndpoint) map[string]v1.Pod {
+func GetStaticPodSpecs(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmapi.APIEndpoint, proxyEnvs []v1.EnvVar) map[string]v1.Pod {
 	// Get the required hostpath mounts
 	mounts := getHostPathVolumesForTheControlPlane(cfg)
+	if proxyEnvs == nil {
+		proxyEnvs = kubeadmutil.GetProxyEnvVars()
+	}
```

Comment:
The code of `GetStaticPodSpecs` is a little tricky here.

It checked if the last arg is nil. If nil, the function will add proxy env. If not nil(including empty), it will not add proxy env. The empty case is used in the test only. https://github.com/kubernetes/kubernetes/pull/118867/files#diff-1937a20cf97bf14e2aea8ae19d61d935536d2c903b49adf4f4502fcfd89a5296R61.

No place will really pass a real env into the function.

```
func GetStaticPodSpecs(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmapi.APIEndpoint) map[string]v1.Pod {
	// Get the required hostpath mounts
	mounts := getHostPathVolumesForTheControlPlane(cfg)
	proxyEnvs := kubeadmutil.GetProxyEnvVars()

```

The test below may just check the last env of the container[0].

Just for nit.

Reply:
this was discussed here: https://github.com/kubernetes/kubernetes/pull/118867#discussion_r1249195215, I am fine with either of them but a little in favor of the current approach (we cannot guarantee extraEnv is always the last env, and actually it's a slice and we cannot inspect easily how many envs the user sets), the original approach is filter the proxy env out, @neolit123 suggest me to remove the dep over the host.

Reply:
Sorry to miss the comments above 😄. 

It seems to be on test purposes to not be affected by host env. 

It is fine.